### ISSUE TITLE
Error Missing data fillup causing unreliable outputs

### ISSUE THREAD

**Issue by QgitM05** [freqtrade/freqtrade#9751]
## Describe your environment

  * Operating system: Windows 64bit
  * Python Version: 3.10
  * Freqtrade Version: 2023.12
  
I've had issues with data downloading where there keep forming large gaps in data. This ruins the output of backtesting/hyperopts. Whatever I've tried, I haven't been able to correctly download the data apparently as I haven't been able to get rid of the gaps.

Also I use kraken as an exchange and on the freqtrade website I've read that there are known issues with kraken data downloading. It also said that you could download all of the data from kraken zipped. I did that and after that I followed the steps converting the data to feather files and then converting it to the timeframes that I want. But when I used that dat I got the following output from backtesting. I got that issue many times before in fact I think I haven't been able to use data once where there weren't any gaps.

I get the following output when trying to backtest.

2024-01-30 20:39:31,103 - freqtrade.data.history.history_utils - INFO - Using indicator startup period: 30 ...
2024-01-30 20:39:31,203 - freqtrade.data.converter.converter - INFO - Missing data fillup for BCH/USDT, 15m: before: 39 - after: 124 - 217.95%
2024-01-30 20:39:31,274 - freqtrade.data.converter.converter - INFO - Missing data fillup for LINK/USDT, 15m: before: 106 - after: 126 - 18.87%
2024-01-30 20:39:31,311 - freqtrade.data.converter.converter - INFO - Missing data fillup for SOL/USDT, 15m: before: 125 - after: 127 - 1.60%
2024-01-30 20:39:31,386 - freqtrade.data.converter.converter - INFO - Missing data fillup for ADA/USDT, 15m: before: 111 - after: 127 - 14.41%
2024-01-30 20:39:31,454 - freqtrade.data.converter.converter - INFO - Missing data fillup for DOT/USDT, 15m: before: 115 - after: 127 - 10.43%
2024-01-30 20:39:31,496 - freqtrade.data.converter.converter - INFO - Missing data fillup for ALGO/USDT, 15m: before: 36 - after: 117 - 225.00%
2024-01-30 20:39:31,498 - freqtrade.optimize.backtesting - INFO - Loading data from 2024-01-27 16:30:00 up to 2024-01-29 00:00:00 (1 days).

It generates the output of the backtest but the output is really unreliable. 

So my obvious question is:
- How do I fix this?

- Also if someone could explain to me what the numbers mean "before: 39 - after: 124" I would very much appreciate that :).

**Comment by xmatthias** [comment]
2024.1 has slight changes to the kraken download method, which you might wanna give a try (delete the trades data first if you don't trust current results).
Using the kraken CSV files will only help you if you need tons of data - for the last few weeks, that data is usually not available.

Now i don't think that'll make a significant difference though ... (read on for details).


Now a few thing to note: 
* Exchange downtimes WILL cause this message - as there were no candles at that time - but will not really influence results (the exchange was gone, anyway).
* no trade periods (pretty common on kraken's USDT pairs - which seem to show very low volume) WILL cause this message as then there's no executed trades to form the candle from.



A small  screenshot for point 2:
Most of the "-" candles have 0 volume - which means 0 executed trades. These candles will "not exist" physically - and will cause these messages - as we can't convert trades to these candles - and therefore store them in a sparse mode (missing the 0 volume candles).
This is not really a problem, as the data fillup generates the candles as they look at the exchange. The messages would be gone if we would run the fillup during generation - but i don't really see the point.

![image](https://github.com/freqtrade/freqtrade/assets/5024695/5ec9caf3-c278-4e34-9742-152a7cfc2ccd)


Other pairs look worse (both screenshots are 15m) - however confirming that the fillup values are roughly correct.

![image](https://github.com/freqtrade/freqtrade/assets/5024695/82f7f3b1-ca3a-48f3-88a8-335a61efb7eb)

Confirming what i said - USDT markets on kraken are very low volume (12k traded in the last 24 h, vs. 785k$ traded on BCH/USD (no T) ... 

![image](https://github.com/freqtrade/freqtrade/assets/5024695/2512db67-8bf2-4ca4-8d23-586a458ed483)


To summarize - i think the candles / messages are correct (and so are the outputs you get from backtesting) - simply because you're on low-volume pairs - which will struggle to fill trades.

**Comment by QgitM05** [comment]
Thank you very much for your quick response and the good explanation. I'm glad to hear that the data downloads were in fact right then and this is 'normal'. Tho this leaves me with 1 question.

I of course understand that there is no guaranteed similar output in live run as the output you got from backtesting. But I've run my strategy live for a while and the results I'm getting are way off compared to when I backtest on that same timerange I've traded live on. Now I don't expected the backtest to deliver the exact same results as my live run but it shows close to none similarities. The results are way different which I find odd. Is there anything I might be doing wrong here?

**Comment by xmatthias** [comment]
well backtesting makes certain [assumptions](https://www.freqtrade.io/en/stable/backtesting/#assumptions-made-by-backtesting) - which may or may not hold true to live trading - but take assumptions of "decent volume" throughout the candles.

on pairs that don't have volume (essentially most /USDT pairs on kraken) - this will usually become worse as backesting may assume an order as "fills" - while in reality it won't (or it will push the price by quite some margin).

The problem here is mainly the spread - which on a pair that doesn't show any trades for hours (see above, BCH) - will be quite monumental, so if you're using a pricing strategy of "other" (or market orders) - you'll not get "candle.open" - but rather "candle.open + 1%" ... which means compared to backtesting, you're already 1% negative "just because".

**Comment by QgitM05** [comment]
Ah thank you I understand. That makes sense. Could this problem be fixed by backtesting with something along the lines of volume>0 in the entry and exit strategies? And would you recommend trading on /EUR or /USD pairs instead of /USDT?

**Comment by xmatthias** [comment]
for entries - yes
for exits - it depends on the exit reason.

Trading on markets with more volume will definitely help - you're going from a volume of 12k/day to 785k/day - more than 50x more volume ... (see above screenshot)

**Comment by stash86** [comment]
Another reading material

https://botacademy.ddns.net/2023/11/11/why-volume-matters/

**Comment by QgitM05** [comment]
Both of you, thank you very much. I learned a lot from this! This will for sure help me
